### PR TITLE
First draft for contribution guide

### PR DESCRIPTION
**Pull request by tlunet**
This is a first draft for a contribution guide (that can be see as a first proposal), with in particular :

- some guidelines for branching
- basic documentation for black and flakeheaven
- a first proposal for naming convention

Of course, some points in the contribution guide can be put to debate (in particular, for the naming convention)

### CODE REVIEW

**Comment by tlunet** [comment]
# An argument (slightly in favor) for camelCase

## First quick definitions

- PascalCase : VariableNameLikeThis
- camelCase : variableNameLikeThis
- snake_case : variable_name_like_this

## Abstract

Even if PEP8 suggests the use of snake_case for variable and function names, the alternative use of camelCase may be better from a practical point of view. This would not be the first time we break PEP8 guidelines (for instance, max line length to 120 in current settings), but there are other arguments.

### Uniform and simplified variable names

- using snake_case allows the following equivalent naming without explicitly ranking those :
```python
tend = 1.0
t_end = 1.0
Tend = 1.0
```
using camelCase gives only one candidate `tEnd` that is simple to read for most people, and would stay consistent with any other time variable naming.

- some variable indicating a number cannot have a short writing in snake_case, for instance `num_nodes` (instead of `nnodes`), while others use the short format (_e.g_ `nvars`). Using camelCase would uniformize the notation for both, while still being clear and concise : `nNodes`, `nVars`.

- mix with acronyms are not very explicit in snake_case. For instance, `mssdc_jac` may be clearer using camelCase and putting acronym at the end : `MSSDCJacobi`. Eventually, one could consider adding underscore to camelCase when it helps separating acronyms, for instance : `useJacobian_MSSDC` (sidenote : that's what I thought `mssdc_jac` was when reading it first :sweat_smile:).

The main point here is that for the applications we are considering (with a lot of `n[blabla]` and other similar parameters), camelCase is probably more efficient and consistent on the long run. And there is for now rare (if none) situations where variable names in pySDC are harder to read in camelCase than in snake_case (for instance, `ìsIllconditionned` or similar name should probably not be used).

### Reading effectiveness and current trend

camelCase is actually used in Python core libraries (for instance, `logging`) and there is no urge of complying to PEP8 (on the contrary, see https://lwn.net/Articles/877115/). Also, a scientific study comparing both snake_case and camelCase support the idea that, even if snake_case is easier to read in general, camelCase 

> "leads to higher accuracy among all subjects regardless of training, and those trained in camel casing are able to recognize identifiers in the camel case style faster than identifiers in the underscore style".

So yes, using camelCase would require a bit of time for people using snake_case to get use to, but not necessarily long. Furthermore, the use of PascalCase for class names is already quite natural in Python, so it won't be a huge gap to go to object methods and attributes written with camelCase. Furthermore, if used wisely (with eventual help of underscore), it could help reducing the length of a code, so it looks more appealing for a user that don't know already how it works. For instance :

```python
# instantiate controllers in snake_case
controller_mssdc_jac = controller_nonMPI(
    num_procs=num_proc, controller_params=controller_params_jac, description=description_mssdc
)
# instantiate controllers in camelCase
controllerMSSDCJacobi = controllerNonMPI(
    nProc=nProc, controllerParams=controllerParamsJacobi, description=descrMSSDC
)
# Note that removing underscore allows a more explicit naming for controllerMSSDCJacobi without too much more space ...
```

From a short poll I did in the math section in TUHH, 5 persons out of 6 that answered where more in favor of camelCase than snake_case (the only person favoring snake_case beeing @brownbaerchen). I know this is a bit bias, and not very representative, but more detailed research indicate that camelCase is slowly becoming a serious alternative to snake_case, in particular with programming languages like TimeScript (next generation JavaScript) that massively use it (Python almost getting there with PascalCase for class names). 

### Work will have to be done anyway

Currently, pySDC use a mix between snake_case and camelCase, with majority for snake_case. Going to snake_case would implies a lot of changes in all naming. However, this will have to be done already for man class names, and with the whole default parameter rework for v6. So this would be the perfect opportunity to refactor most of pySDC names (let the change be a full one :smile:). And also, many IDEs and editors offer some very simple ways to change a given variable name now ...

So what do you think @pancetta ?

**Comment by pancetta** [comment]
> This is a first draft for a contribution guide (that can be see as a first proposal), with in particular :
> 
> * some guidelines for branching
> * basic documentation for black and flakeheaven
> * a first proposal for naming convention
> 
> Of course, some points in the contribution guide can be put to debate (in particular, for the naming convention)

I really appreciate your contribution, thanks s lot. Before you write more markdown files, though: does that play along with the ReStructuredText files used so far? Esp. for the website?

**Comment by pancetta** [comment]
I’m totally fine with camelCase!

**Comment by tlunet** [comment]
> > This is a first draft for a contribution guide (that can be see as a first proposal), with in particular :
> > 
> > * some guidelines for branching
> > * basic documentation for black and flakeheaven
> > * a first proposal for naming convention
> > 
> > Of course, some points in the contribution guide can be put to debate (in particular, for the naming convention)
> 
> I really appreciate your contribution, thanks s lot. Before you write more markdown files, though: does that play along with the ReStructuredText files used so far? Esp. for the website?

That's an answer I don't have the courage to look for on a Sunday evening :sweat_smile:

**Comment by tlunet** [comment]
> > > This is a first draft for a contribution guide (that can be see as a first proposal), with in particular :
> > > 
> > > * some guidelines for branching
> > > * basic documentation for black and flakeheaven
> > > * a first proposal for naming convention
> > > 
> > > Of course, some points in the contribution guide can be put to debate (in particular, for the naming convention)
> > 
> > 
> > I really appreciate your contribution, thanks s lot. Before you write more markdown files, though: does that play along with the ReStructuredText files used so far? Esp. for the website?
> 
> That's an answer I don't have the courage to look for on a Sunday evening sweat_smile

Finally got the courage to look into it. Markdown are naturally not included in the website. I tried using rst files quickly.
But having links in rst files that automatically include rst files in sphinx doc is also a pain ! Also, you don't have emoji as in markdown, and the whole rst syntax is way heavier than markdown. 

So the simplest solution I found : include a direct link to the github repo page : make sense since potential contributors will mostly look at the repo than the website. I will just correct this link once the page is included to master. Also, this will be much more easier to include links to code and other pages with graphs (and other stuff) when building the `04_custom_implementations.md` page.

**Comment by brownbaerchen** [comment]
You write markdown files are not included in the website naturally, but maybe there is a way we can include both existing rst files and new markdown files in the website? A quick google suggested the use of `mdinclude` to include markdown into rst, which is processed by sphinx. However this already feels like a huge mess.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+1. [Pull Requestion recommendations](./docs/contrib/01_pull_requests.md)
+2. [Continuous Integration](./docs/contrib/02_continuous_integration.md)
+3. [Naming Conventions](./docs/contrib/03_naming_conventions.md)
+4. [Custom Implementations](./docs/contrib/04_custom_implementations.md)
+
+Developments on the `pySDC` code use the classical approach of forks and pull requests from Github.
+There is an [extended documentation](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/getting-started/about-collaborative-development-models) on this aspect (that you can skip if you are already used to it). In addition, some recommendations for pull requests are given [here](./docs/contrib/01_pull_requests.md).
+
+Additionnaly, a _few_ rules are set to enforce code readability, consistency and reliability. Some of them are automatically tested with each commit, and summarized in the page on [continuous integration (CI)](./docs/contrib/02_continuous_integration.md).
```
Comment (brownbaerchen) [inline CONTRIBUTING.md:11]: There is one excess n and one missing l in "Additionnaly". I am sure there are spellcheckers for VScode, but it also really easy in vim. You can just write `:set spell` to activate spell checking. When there is a word not in vims dictionary, you can easily add it by writing `:spell pySDC` for instance.
Sorry for nitpicking, but I think we should spell check text only files.

**Comment by brownbaerchen** [comment]
I really like what you've written! I think it makes pySDC appear quite a bit more accessible.

**Comment by tlunet** [comment]
> You write markdown files are not included in the website naturally, but maybe there is a way we can include both existing rst files and new markdown files in the website? A quick google suggested the use of `mdinclude` to include markdown into rst, which is processed by sphinx. However this already feels like a huge mess.

That's one thing I tried indeed. You can do that for sphinx to include Markdown file, but then the GitHub renderer will not interpret this as a link. So the documentation will be available in the website generated by sphinx, but not in the repo. The main problem here is the use of the main README.rst file to build the website. However, I wonder if we put the README in mardown format, and then use `mdinclude` in the index.rst file of the documentation, does it include this file and all linked README file (not sure though ...)

**Comment by tlunet** [comment]
> I really like what you've written! I think it makes pySDC appear quite a bit more accessible.

Even the camelCase thing ? :wink:

**Comment by brownbaerchen** [comment]
> > You write markdown files are not included in the website naturally, but maybe there is a way we can include both existing rst files and new markdown files in the website? A quick google suggested the use of `mdinclude` to include markdown into rst, which is processed by sphinx. However this already feels like a huge mess.
> 
> That's one thing I tried indeed. You can do that for sphinx to include Markdown file, but then the GitHub renderer will not interpret this as a link. So the documentation will be available in the website generated by sphinx, but not in the repo. The main problem here is the use of the main README.rst file to build the website. However, I wonder if we put the README in mardown format, and then use `mdinclude` in the index.rst file of the documentation, does it include this file and all linked README file (not sure though ...)

Oh man, sphinx and GitHub are just such a mess together... D:

File: CONTRIBUTING.md (hunk 1 above)
Comment (tlunet) [reply CONTRIBUTING.md]: Done for all file, and you're right ! Quite easy to activate a spell-checking extension in VSCode, and it won't hurt my approximate English writing :sweat_smile:

**Comment by pancetta** [comment]
My 2 cents: I'm ok with having markdown files in the docs, as long as they're clearly targeted toward Github users/developers. My original thought was that I don't want to write documentation twice (website + Github), but in order to get things done, we could just have `rst` files where necessary and possible and `md` files for the lazy.. I mean.. where it's not straightforward.

**Comment by tlunet** [comment]
grafik

Do I see here some love-hate relationship with camelCase @brownbaerchen :smile: ?

**Comment by brownbaerchen** [comment]
> grafik
> 
> Do I see here some love-hate relationship with camelCase @brownbaerchen 😄 ?

Well, the hate is directed towards camel case, but the love was meant for you ;)

**Comment by tlunet** [comment]
> My 2 cents: I'm ok with having markdown files in the docs, as long as they're clearly targeted toward Github users/developers. My original thought was that I don't want to write documentation twice (website + Github), but in order to get things done, we could just have `rst` files where necessary and possible and `md` files for the lazy.. I mean.. where it's not straightforward.

Then let's do this for the contribution guide : host it on github with markdown files, and put a full link in the README.rst such that it will be integrated as well in the documentation website.

**Comment by brownbaerchen** [comment]
That makes sense. Since the contributing happens on GitHub, it should be prioritised for the corresponding documentation imo. I was really hoping you would find a long term solution for documentation of projects both on GitHub and on the website, though. :D

**Comment by tlunet** [comment]
> That makes sense. Since the contributing happens on GitHub, it should be prioritised for the corresponding documentation imo. I was really hoping you would find a long term solution for documentation of projects both on GitHub and on the website, though. :D

Created #250 for that and an associated milestone :smiley:

**Comment by tlunet** [comment]
I've messed up thing very bad (accidentally merged v6 into contrib through an other branch ...). I've restarted a new contrib-2 branch from start using cherry-picks, and opened a new PR #257.

**Pull request closed without merging**